src/peer_transport/handler.ts

```diff
@@ -0,0 +1,151 @@
+import { TimeoutController } from 'timeout-abort-controller'
+import { readCandidatesUntilConnected } from './util.js'
+import * as pb from './pb/index.js'
+import { abortableDuplex } from 'abortable-iterator'
+import { logger } from '@libp2p/logger'
+import type { Stream } from '@libp2p/interface-connection'
+import type { StreamMuxerFactory } from '@libp2p/interface-stream-muxer'
+import { DataChannelMuxerFactory } from '../muxer.js'
+
+const DEFAULT_TIMEOUT = 30 * 1000
+
+const log = logger('libp2p:webrtc:peer')
+
+export type IncomingStreamOpts = { rtcConfiguration?: RTCConfiguration } & IncomingStreamData
+
+export async function handleIncomingStream ({ rtcConfiguration, stream: rawStream }: IncomingStreamOpts): Promise<[RTCPeerConnection, StreamMuxerFactory]> {
+  const timeoutController = new TimeoutController(DEFAULT_TIMEOUT)
+  const signal = timeoutController.signal
+  const stream = pbStream(abortableDuplex(rawStream, timeoutController.signal)).pb(pb.Message)
+  const pc = new RTCPeerConnection(rtcConfiguration)
+  const muxerFactory = new DataChannelMuxerFactory(pc)
+
+  const connectedPromise: DeferredPromise<void> = pDefer()
+  signal.onabort = () => { connectedPromise.reject() }
+  // candidate callbacks
+  pc.onicecandidate = ({ candidate }) => {
+    stream.write({
+      type: pb.Message.Type.ICE_CANDIDATE,
+      data: (candidate != null) ? JSON.stringify(candidate.toJSON()) : ''
+    })
+  }
+
+  // setup callback for peerconnection state change
+  pc.onconnectionstatechange = (_) => {
+    log.trace('receiver peerConnectionState state: ', pc.connectionState)
+    switch (pc.connectionState) {
+      case 'connected':
+        connectedPromise.resolve()
+        break
+      case 'failed':
+      case 'disconnected':
+      case 'closed':
+        connectedPromise.reject()
+        break
+      default:
+        break
+    }
+  }
+  // read an SDP offer
+  const pbOffer = await stream.read()
+  if (pbOffer.type !== pb.Message.Type.SDP_OFFER) {
+    // TODO: Find better way to print undefined without linter complaining
+    throw new Error(`expected message type SDP_OFFER, received: ${pbOffer.type ?? 'undefined'} `)
+  }
+  const offer = new RTCSessionDescription({
+    type: 'offer',
+    sdp: pbOffer.data
+  })
+
+  await pc.setRemoteDescription(offer).catch(err => {
+    log.error('could not execute setRemoteDescription', err)
+    throw new Error('Failed to set remoteDescription')
+  })
+
+  // create and write an SDP answer
+  const answer = await pc.createAnswer()
+  // write the answer to the remote
+  stream.write({ type: pb.Message.Type.SDP_ANSWER, data: answer.sdp })
+
+  await pc.setLocalDescription(answer).catch(err => {
+    log.error('could not execute setLocalDescription', err)
+    throw new Error('Failed to set localDescription')
+  })
+
+  // wait until candidates are connected
+  await readCandidatesUntilConnected(connectedPromise, pc, stream)
+  // close the dummy channel
```

Comment:
Where do we close the channel?

Reply:
Sorry, forgot to delete this comment from a previous implementation. Earlier, the handler was creating a channel, and was therefore responsible for closing it. Now the channel is created by the initiator. The purpose of creating a dummy channel is to add a WebRTC component for which candidates should be gathered.